### ISSUE TITLE
Qwen4Exp (Qwen3.8-Flash-Next) cannot emit DFlash aux hidden states: DFlash/DSpark drafters cannot be served

### ISSUE THREAD

**Issue by seanphan** [sgl-project/sglang#38589]
### Summary

`Qwen4Exp` (Qwen3.8-Flash-Next) cannot emit DFlash aux hidden states, so no DFlash or DSpark drafter can be served on this target. `set_dflash_layers_to_capture` is inherited and appears to succeed, but the captures are never collected and the return slot they would use is already occupied by the hyper-connection stream. This is the same class of gap that `glm5_next` had before #36708, and the fix has a direct precedent in that model.

Checked on `main` and on the pinned cookbook image `lmsysorg/sglang:qwen38flashnext` (sglang `d91c3682b0b429e4c70df63cd57f819588ce29b0`, 2026-08-16). Both behave identically. Line numbers below are `main`; the pinned image is +12 in `models/qwen4_exp.py`.

This is a read-only report — I am not opening a PR. Happy to test a fix on real hardware, see the end.

### What works

`configure_aux_hidden_state_capture` (`python/sglang/srt/model_executor/model_runner_components/attention_backend_setup.py:47-66`) calls `model.set_dflash_layers_to_capture(...)`. For `Qwen4ExpForConditionalGeneration` that resolves `models/qwen3_vl.py:1542` -> `models/qwen3_5.py:1736`, which sets `_is_layer_to_capture = True` on the selected layers. No error is raised at any point, so the configuration looks like it took effect.

### What is broken

All three in `python/sglang/srt/models/qwen4_exp.py`:

1. **`layer_communicator` is deleted.** `Qwen4ExpLayerExtensionMixin._init_qwen4_exp_layer_extensions:1222-1228` `delattr`s `input_layernorm`, `post_attention_layernorm` and `layer_communicator`. `layer_communicator.prepare_attn_and_capture_last_layer_outputs` is the call that actually appends captures in the parent (`models/qwen3_5.py:1021,1461`), so the collection mechanism is removed.

2. **Neither decoder layer consumes `captured_last_layer_outputs`.** `Qwen4ExpAttentionDecoderLayer.forward:1535-1561` and `Qwen4ExpLinearDecoderLayer.forward:1392-1414` fully override `forward` and take the kwarg into `**kwargs` without reading it. In the pinned image the identifier `captured_last_layer_outputs` occurs exactly once in the whole file — at the call site, `:1658`. `aux_hidden_states` built at `Qwen4ExpModel.forward:1632` can therefore never be non-empty.

3. **The second return slot collides with the hyper-connection stream.** `Qwen4ExpModel.forward:1655-1662`:

```python
hc_hidden_states = hidden_states
hidden_states, _ = self.hyper_connection_mixer.mix(hidden_states)
if not forward_batch.forward_mode.is_idle():
    return hidden_states, hc_hidden_states     # every real forward takes this
if len(aux_hidden_states) == 0:
    return hidden_states
return hidden_states, aux_hidden_states        # reachable only when idle
```

`Qwen4ExpForConditionalGeneration.forward:1735-1738` then sets `output.hidden_states = hc_hidden_states`. So even if (1) and (2) were fixed, the aux tensors would have nowhere to go: the tuple's second element is unconditionally the HC stream on every non-idle forward.

### Consequence

`DFlashWorkerV2` reads `logits_output.hidden_states` as its target context features (`python/sglang/srt/speculative/dflash_worker_v2.py:1947,2341`) and projects them through `DFlashDraftModel.project_target_hidden` (`python/sglang/srt/models/dflash.py:659-676`). On this target it receives the HC-flattened `hc_count(4) x hidden_size(2560) = 10240`-wide last-layer stream rather than a concatenation of per-layer taps.

A draft checkpoint whose `fc` in_features is not exactly 10240 fails with the `project_target_hidden` feature-dim `ValueError`. One that happens to be 10240 wide is worse: it loads and drafts from semantically wrong features with no error at all.

Consistent with the above, `grep -c dflash python/sglang/srt/models/qwen4_exp.py` returns `0` on both `main` and the pinned image.

### Precedent for the fix

`models/glm5_next.py` solved exactly this for its own mHC model in #36708:

- `_prepare_aux_hidden_state:953-962` applies `hc_contract(aux_hidden_state, self.config.hc_mult)` when `self.dflash_capture and self.config.mhc`, contracting the widened stream back to `hidden_size`.
- `set_dflash_layers_to_capture:1242-1252` sets `self.model.dflash_capture = True` alongside the per-layer flags.
- The collection path in `:1016-1034,1074-1076` returns `aux_hidden_states` rather than a competing stream.

A `qwen4_exp` fix would need the analogous three pieces: an HC-contracting aux hook, a `dflash_capture` flag set by `set_dflash_layers_to_capture`, and a resolution of the `(hidden_states, hc_hidden_states)` vs `(hidden_states, aux_hidden_states)` return collision at `:1655-1662` (a third element, or carrying aux on the model like `last_hc_hidden_states` already is at `:1685-1711`).

### Related, and a possible second issue

The cookbook page for this model currently offers only `--speculative-algorithm NEXTN`, which matches the state of the code — so the docs are not wrong today. But if a DFlash/DSpark path is ever enabled here, note that `DFlashWorkerV2` assumes the `1+N` block layout (`dflash_worker_v2.py:268` and `:2183-2193`, `hidden_states.view(bs, block_size, -1)[:, 1:, :]`, position 0 being the anchor and its own output discarded). Checkpoints trained with the anchor-as-first-prediction layout — where every position in the block predicts — would be shifted by one position **silently**, with no error and no obvious signature beyond poor acceptance. A `sample_from_anchor` check on the DFlash path, matching how `is_nemotron_35_draft_config` already reads that field (`speculative/dflash_utils.py:503-506`), would turn that into a loud failure. Happy to split this into its own issue if you would prefer.

### Motivation / offer

We have a DFlash/DSpark-family drafter trained against this target that we cannot serve on SGLang because of the above. Its config is otherwise a clean fit for the existing loader — `architectures: ["Qwen3DSparkModel"]` is a registered EntryClass (`models/dspark.py:866,884`), and its `fc` in_features equals `len(target_layer_ids) * hidden_size`, exactly what `DFlashDraftModel` computes.

We run Qwen3.8-Flash-Next NVFP4 on 2x DGX Spark (GB10, TP2) and are willing to test a patch end to end on that hardware and report back. I have deliberately not opened a PR, since the right shape of the return-slot fix is a call for the maintainers of this model.

**Comment by Leslie360** [comment]
Opened #38642 with a fix following the #36708 GLM-5.3-Flash precedent — all three diagnosed breakages are addressed inside `qwen4_exp.py` (model-level capture loop; aux states own the second return slot when capture is active; HC stream no longer clobbers packed aux states). CPU-level contract tests included; no hardware verification on my side yet — @seanphan, since you offered to test, the checklist in the PR body should cover what to run. Thanks for the careful write-up; the three-point diagnosis mapped 1:1 onto the fix.
